halo2/src/recursion/endoscale.rs

```diff
@@ -0,0 +1,102 @@
+//! Gadget for endoscaling.
+
+use ff::PrimeFieldBits;
+use halo2_proofs::{
+    circuit::{AssignedCell, Layouter},
+    plonk::Error,
+};
+use pasta_curves::arithmetic::CurveAffine;
+
+mod chip;
+pub mod primitive;
+
+/// Instructions to map bitstrings to and from endoscalars.
+pub trait EndoscaleInstructions<C: CurveAffine>
+where
+    C::Base: PrimeFieldBits,
+{
+    /// A bitstring to be used in endoscaling.
+    type Bitstring: std::fmt::Debug + Clone;
+    /// The maximum number of bits that can be represented by [`Self::Bitstring`].
+    const MAX_BITSTRING_LENGTH: usize;
+
+    /// Witnesses a slice of bools as a vector of [`Self::Bitstring`]s.
+    fn witness_bitstring(bits: &[bool]) -> Vec<Self::Bitstring>;
```

Comment:
This doesn't take a layouter, so it prevents us from asserting that `EndoscaleInstructions::Bitstring` is always correctly constrained. We should pass a layouter argument in here, and then even if the chip doesn't use it, the other parts of the instruction API (as used externally by gadgets) can transitively assume that every `Bitstring` is correctly constrained (even if the chips do so lazily).